3rdparty/dlmalloc/CMakeLists.txt

```diff
@@ -3,27 +3,36 @@
 
 add_enclave_library(oedlmalloc OBJECT allocator.c)
 
-enclave_link_libraries(oedlmalloc PRIVATE oe_includes oelibc_includes)
+enclave_link_libraries(oedlmalloc PRIVATE oe_includes)
 
 if (OE_TRUSTZONE)
-    enclave_link_libraries(oedlmalloc PUBLIC oelibutee_includes)
-    set(TEE_C_FLAGS ${OE_TZ_TA_C_FLAGS})
-else()
-    set(TEE_C_FLAGS "")
-endif()
+  enclave_link_libraries(oedlmalloc PUBLIC oelibutee_includes)
+  set(TEE_C_FLAGS ${OE_TZ_TA_C_FLAGS})
+else ()
+  set(TEE_C_FLAGS "")
+endif ()
 
-enclave_compile_options(oedlmalloc PRIVATE
+enclave_compile_options(
+  oedlmalloc
+  PRIVATE
   -ftls-model=local-exec
   -nostdinc
   -fPIE
   -ffreestanding
   -fvisibility=hidden
   ${TEE_C_FLAGS})
 
+enclave_include_directories(oedlmalloc PRIVATE
+                            ${PROJECT_SOURCE_DIR}/include/openenclave/corelibc)
```

Comment:
What is the benefit of including corelibc with OE_NEED_STDC_NAMES rather than including oelibc_includes?

Reply:
This allows oecore to be self-contained. Similar to we build memset/memmove/memcpy in oecore using corelibc with OE_NEED_STDC_NAMES.

Reply:
As far a I am aware, oelibc and corelibc with stdc names provide only function declarations right? 
If you compile allocator.c with oelibc and with corelibc would the generated object change?
Do you see any difference in the .o file?

Reply:
Without the changes of the PR, the oecore has a dependency on oelibc. So, the question is more about whether we want to make oecore depend on oelibc, or make it self-contained. I feel the former is kind of conflicting with the reason for corelibc.

Reply:
> This allows oecore to be self-contained. Similar to we build memset/memmove/memcpy in oecore using corelibc with OE_NEED_STDC_NAMES.

These are 4 functions defined as part of a `freestanding` environment. Even `oecore` cannot use a `oe_` prefix for these funcitions.

https://github.com/openenclave/openenclave/blob/7bc2e60523313e360b557bf30fccd293f05563c1/include/openenclave/corelibc/string.h#L25-L29

Reply:
I am fine with these changes. I view an allocator as something that is not tightly coupled with oecore. An allocator can be written in C++ (like snmalloc, tcmalloc, jemalloc) and still be made to work within enclaves. I don't think the requirement `an allocator ought to use corelibc (oecore)`, though seems pure, is feasible in practice considering the allocators that are in existence.

Reply:
Yes, this is something I was looking into as well (and confuses me a little bit). The current behavior is that `oecore` directly includes the sources of these four functions, and `oelibc` includes them as well (without this PR). Now when an enclave app use any of these functions, the app can link to either the object in `oecore` or `oelibc` (depend on their order during the link-time). This is kind of weird to me so in this PR I remove the duplicates from `oelibc`.

We can also go the other direction; i.e., make `oecore` depend on `oelibc` and remove the duplicates from `oecore`. But again, this makes me wonder why we need corelibc if we go for this direction.

Reply:
I also agree that we don't need to make this as a hard requirement to any plugged allocator. But I think it makes sense for the default allocator (dlmalloc) to use corelibc instead of libc.

Reply:
Those 4 functions are special. GCC expects a `freestanding` environment to provide those functions; even when compiling the kernel GCC will still emit calls to those functions.

I'd like for us to be able to change the default allocator; say tomorrow we could choose to go with snmalloc or tcmalloc as the default allocator - couple of teams have already found the our default allocator (dlmalloc) to be non performant.
Thats why, I feel the default allocator itself ought to be compiled as if any other allocator is compiled. So that we may easily replace it at a future date.

Having said all that, I am fine with the changes in this PR. But it is important to be cognizant of the implications in regards to pluggable allocators, marshalling strategies etc.

Reply:
Thanks for the inputs. I'll hold off this PR before having more clear idea of the benefits/high-level goals.